Expose several new Metrics

zentol opened the pull request:
This PR exposes several new metrics in Flink.

The names, scope and respective performance impact (outside of the actual metric itself) are listed blow:

Metric | Scope | Performance Impact
-------|-------|-------------------
currentLowWatermark | Task | None
lastCheckpointSize | Task | StreamTaskStateList.getStateSize() for every checkpoint
numBytes(In/Out)(Local/Remote) | Task |  Buffer.getSize() for every buffer
numRecords(In/Out) | Operator | Streaming: None / Batch: some additional null checks
numSplitsProcessed | Operator | None

The currentLowWatermark metric simply exposes the last emitted Watermark in the StreamInputProcessor using a Gauge.

The lastCheckpointSize metric is gathered in the StreamTask.performCheckpoint method. For every checkpoint made the size is computed and stored in a field, which is exposed with a Gauge.

The numSplitsProcessed metric is a simple counter in the DataSourceTask/FileSourceFunction.

The numBytes(In/Out)(Local/Remote) metric touches the 2nd most files. It replaces the previous numBytes(In/Out) metric which was exposed in the Deserializers. It is now gathered directly in the InputChannels, adding the size of read buffers to a Counter. It is thus also more accurate than the previous version, since that one actually measured you many bytes were deserialized.

The numRecords(In/Out) metric was extended to measure per operator instead of task. This includes chained operators. The information is no longer gathered in the deserializers but the operator itself. This was generally accomplished by wrapping in- and output in Counting(Output/Collector/Iterator/...) classes. Below are more specific details for both Streaming and Batch as to how these were implemented.

### Streaming
* numRecordsIn is measured in the StreamInputProcesser or the ChainingOutput class.
* numRecordsOut is measured by wrapping the output in a CountingOutput, set in the AbstractStreamOperator.

The StreamSource class was slightly modified, adding a second run() method without a collector argument. Using this method the StreamSource now uses the same output that is set in the AbstractStreamOperator. The old method was kept in place so as not to change tests.

The StreamIterationTail class needed some extra love as well, since the contained RecordPusher did not follow the StreamOperator lifecycle. On this operator setup() was never called, leading to an NPE since the MetricGroup was never set. In order to be able to pass an Output object into setup() the RecordPusher's internal logic was moved into a new IterationTailOutput class. This slightly modified the exception behaviour; InterruptedExceptions are now longer simply forwarded, but wrapped in a RuntimeException instead.

### Batch
* numRecordsIn is measured in each Driver class separately. In cases where an iterator is passed into the UDF the underlying MutableObjectIterator was wrapped in a CountingMutableObjectIterator. This should ensure that no input record is counted multiple times.
* numRecordsOut is measured by wrapping the collector in a CountingCollector, set in each Drivers prepare() or run() method.

Note that the following drivers were nto changed: (Join/CoGroup)WithSolutionSet(First/Second)Driver.
I coiuldn't figure out a way to not count records multiple times, specifically those residing in the HashTable, so I ignored them.

fhueske commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamIterationTail.java:
```diff
@@ -56,39 +57,60 @@ public void init() throws Exception {
+		public void collect(StreamRecord<IN> record) {
+			try {
+				if (shouldWait) {
+						dataChannel.offer(record, iterationWaitTime, TimeUnit.MILLISECONDS);
```
Indention

fhueske commented on flink-core/src/main/java/org/apache/flink/metrics/groups/IOMetricGroup.java:
```diff
@@ -26,33 +26,27 @@
 
-	public Counter getRecordsInCounter() {
-		return numRecordsIn;
+	public Counter getNumBytesInLocal() {
```
rename to `getNumBytesInLocalCounter` to be consistent with previous methods?

fhueske commented on flink-core/src/main/java/org/apache/flink/metrics/groups/IOMetricGroup.java:
```diff
@@ -26,33 +26,27 @@
 
-	public Counter getRecordsOutCounter() {
-		return numRecordsOut;
+	public Counter getNumBytesInRemote() {
```
rename to `getNumBytesInRemoteCounter`?

fhueske commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/serialization/AdaptiveSpanningRecordDeserializer.java:
```diff
@@ -181,8 +168,6 @@ public void setReporter(AccumulatorRegistry.Reporter reporter) {
 
 	@Override
 	public void instantiateMetrics(IOMetricGroup metrics) {
```
Do we need this method in `RecordDeserializer`? Both implementations of `RecordDeserializer` implement it as empty method.

fhueske commented on flink-runtime/src/main/java/org/apache/flink/runtime/operators/DataSourceTask.java:
```diff
@@ -165,6 +172,7 @@ public void invoke() throws Exception {
 					// close. We close here such that a regular close throwing an exception marks a task as failed.
 					format.close();
 				}
+				splitCounter.inc();
```
Should we make the semantic of this counter more explicit and call it `completedSplitCounter`?

fhueske commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/FileSourceFunction.java:
```diff
@@ -123,9 +127,11 @@ public void run(SourceContext<OUT> ctx) throws Exception {
 			OUT nextElement = serializer.createInstance();
 			nextElement =  format.nextRecord(nextElement);
 			if (nextElement == null && splitIterator.hasNext()) {
+				splitCounter.inc();
```
In `DataSourceTask` the counter is incremented when a split is closed. We should sync both classes.

zentol replied:
the behaviour in both classed is the same. If the format returns null it is considered empty (==completed), hence we increase the counter.

well the FIleSourceFunction doesn't properly call close() once the split was processed.

zentol replied:
I'll remove the splitCounter metric for now, and add it back in once #2100 is merged.

fhueske commented:
Hi @zentol, didn't find anything crucial. Just a few minor comments.

zentol commented:
@fhueske I've addressed your comments. I also removed the setMetricGroup method in the ReaderBase class (all implementations were empty), add added the watermark metric to the StreamTwoInputProcessor as well.

fhueske commented:
Thanks for the update @zentol. Good to merge, IMO
